Fraud & Risk tools improvements from feedback

eduardoumpierre opened the pull request:
Fixes Automattic/woocommerce-payments-server#3181

#### Changes proposed in this Pull Request

Layout changes:
- Adjust advanced fraud settings container width
- Adjust advanced fraud rule cards spacings
- Adjust discoverability banner spacings

Improvements:
- Add "Held for review and manually canceled" fraud meta box
- Add "Back to Payments Settings" and "Save changes" buttons to the advanced settings footer
- Refetch "Risk review" data after approving or canceling an on-hold order

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Pre-requisites**
* Checkout to this branch `fix/3181-demo-feedback`
* Update the version on `./woocommerce-payments.php` to be at least `5.7.0`
* Enable the fraud protection feature flag
  * If this is the first time, run `npm run wp option add wcpay_fraud_protection_settings_active 1`
  * If you already have it set, run `npm run wp option update wcpay_fraud_protection_settings_active 1`
* Build the client `npm run start`
* Listen to Stripe webhooks `npm run listen` – or equivalent

**Testing**
* Go to **Payments > Settings** and scroll to the **Fraud protection** section.
* Select the **Advanced** protection level and click on "Configure"
* Scroll to the bottom of the page
  * You should see the "Back to Payments Settings" and "Save changes" buttons
* Scroll to the **Purchase Price Threshold** rule
* Enable the filtering by clicking on "Screen transactions for abnormal purchase prices"
* Set **1000** to the "Minimum purchase price" input
* Disable the "Block Payment" check (If it was checked)
* Save the changes
* Open an incognito browser instance, go to your shop, and try to buy something – it should succeed
* Open **WooCommerce > Orders** and click on the first item – it should be the order that you've just created
* Click on "Review payment"

<img width="290" alt="image" src="https://user-images.githubusercontent.com/16882226/228725287-970e76f3-2807-4ad8-a372-a3c22c2bf5a6.png">

* Click on "Block transaction"
* Go back to the order details page
* You should see the new fraud meta box:

<img width="290" alt="image" src="https://user-images.githubusercontent.com/16882226/228725488-c4b81b29-aefb-460b-aea3-b8e3d5a5f221.png">

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

eduardoumpierre committed:
Add fraud meta box to manually blocked orders

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5900` or branch name `fix/3181-demo-feedback` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/3181-demo-feedback) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c3646b7b9860c44d6eadceb46e6b016d3556feea
- Build time: 2023-03-30 15:10:36 UTC

_Note: the build is updated when a new commit is pushed to this PR._

eduardoumpierre committed:
Refetch risk review data after approving or blocking an on hold transaction

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into fix/3181-demo-feedback

eduardoumpierre committed:
Adjust advanced fraud settings container width to match the design

eduardoumpierre committed:
Add buttons to the advanced fraud settings footer

eduardoumpierre committed:
Fix advanced fraud settings unit tests

eduardoumpierre committed:
Adjust fraud protection rule card spacings

eduardoumpierre committed:
Adjust fraud risk tools banner style to match the design

eduardoumpierre committed:
Fix mark payment capture canceled unit tests

eduardoumpierre committed:
Add changelog entry

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into fix/3181-demo-feedback

tpaksu committed:
Merge branch 'develop' into fix/3181-demo-feedback

tpaksu reviewed (approved):
Tested the instructions and it works, saw the buttons below the advanced configuration form, both worked well, when I clicked one of them, the other one got busy and disabled too, also the metabox update worked well. Thanks for fixing the issues from the demo. LGTM.

tpaksu committed:
Merge branch 'develop' into fix/3181-demo-feedback

tpaksu committed:
Fix HoverTooltip import and usage

jessepearson commented on includes/class-wc-payments-order-service.php:
```diff
@@ -231,25 +234,28 @@ public function mark_payment_capture_expired( $order, $intent_id, $intent_status
+		 */
+		if ( isset( $intent_data['fraud_outcome'] ) && Rule::is_valid_fraud_outcome_status( $intent_data['fraud_outcome'] ) ) {
+			$this->set_fraud_outcome_status_for_order( $order, $intent_data['fraud_outcome'] );
+			$this->set_fraud_meta_box_type_for_order( $order, Fraud_Meta_Box_Type::REVIEW_BLOCKED );
```
This is making the assumption that the intent was cancelled because it was held for review. I believe a check should be added to see if the initial `fraud_outcome` for the order was `review` before setting it to blocked.

eduardoumpierre replied:
That's a good idea! The check was added on [c383b20](https://github.com/Automattic/woocommerce-payments/pull/5900/commits/c383b200be22ea794cf4d3465c337e4f97581b6c).

jessepearson commented on includes/class-wc-payments-order-service.php:
```diff
@@ -1364,7 +1370,7 @@ private function update_order_status( $order, $order_status, $intent_id = '' ) {
 	 * @return array The data we need to continue processing.
 	 */
-	private function get_intent_data( $intent ): array {
+	public function get_intent_data( $intent ): array {
```
This method is a hack due to intents can be objects or arrays and is meant to be private for that reason, it shouldn't be used outside of this class. When I added it previously I should have been clearer in the docblock.

eduardoumpierre replied:
Addressed on [c383b20](https://github.com/Automattic/woocommerce-payments/pull/5900/commits/c383b200be22ea794cf4d3465c337e4f97581b6c).

jessepearson commented on includes/class-wc-payments-order-service.php:
```diff
@@ -231,25 +234,28 @@ public function mark_payment_capture_expired( $order, $intent_id, $intent_status
-			return;
-		}
-
+	public function mark_payment_capture_cancelled( $order, $intent_data ) {
```
If this method is now only called through the public method `update_order_status_from_intent`, then it should be updated to be private and moved down to the private methods. It was only public before because it was being called directly. When I did the refactor, I only updated the methods that were already in the `update_order_status_from_intent` in the gateway class.

eduardoumpierre replied:
Thanks for catching those details. It was addressed on [c383b20](https://github.com/Automattic/woocommerce-payments/pull/5900/commits/c383b200be22ea794cf4d3465c337e4f97581b6c).

jessepearson commented on tests/unit/test-class-wc-payments-order-service.php:
```diff
@@ -636,8 +636,10 @@ public function test_mark_payment_capture_cancelled() {
+
 		// Act: Attempt to mark the payment/order expired/cancelled.
-		$this->order_service->mark_payment_capture_cancelled( $this->order, $intent->get_id(), $intent->get_status() );
+		$this->order_service->mark_payment_capture_cancelled( $this->order, $intent_data );
```
The test can call `update_order_status_from_intent` here with the `$intent`, and it should handle things from there. This will allow `get_intent_data` to remain private.

eduardoumpierre replied:
Oh, that makes sense. Thanks for the tip! It was addressed on [c383b20](https://github.com/Automattic/woocommerce-payments/pull/5900/commits/c383b200be22ea794cf4d3465c337e4f97581b6c).

jessepearson commented on tests/unit/test-class-wc-payments-order-service.php:
```diff
@@ -654,7 +656,7 @@ public function test_mark_payment_capture_cancelled() {
 
 		// Assert: Applying the same data multiple times does not cause duplicate actions.
-		$this->order_service->mark_payment_capture_cancelled( $this->order, $intent->get_id(), $intent->get_status() );
+		$this->order_service->mark_payment_capture_cancelled( $this->order, $intent_data );
```
The test can call `update_order_status_from_intent` here with the `$intent`.

eduardoumpierre replied:
Addressed on [c383b20](https://github.com/Automattic/woocommerce-payments/pull/5900/commits/c383b200be22ea794cf4d3465c337e4f97581b6c).

jessepearson reviewed (changes requested):
Since this was already reviewed and approved, I only addressed the meta box code brought up in Slack: p1680175832897209-slack-C01B8KNUYSW

There are some changes that should be addressed before this is merged, so that we can make less assumptions about why an intent is cancelled and continue to clean up the code.

eduardoumpierre committed:
Improve code based on review

Changed methods that aren't used outside the class to be private.

Updated the tests to match the call the correct method.

Added a check to only mark as "review_block" orders that were previously
marked as "review"

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into fix/3181-demo-feedback

eduardoumpierre committed:
Add unit tests for advanced fraud risk tools settings

eduardoumpierre committed:
Update copy from fraud protection block toggle

eduardoumpierre committed:
Merge remote-tracking branch 'origin/develop' into fix/3181-demo-feedback

eduardoumpierre committed:
Fix unit test code style

jessepearson reviewed (approved):
Thanks for addressing the items I commented on @eduardoumpierre .

It was a good catch moving `mark_payment_capture_cancelled` into `update_order_status_from_intent`, this really helps us clean up the code and move things to a central location for order handling.
